plugin/pkg/replacer/replacer.go

```diff
@@ -41,168 +56,247 @@ var labels = []string{
-		}
-	}
-
-	b.WriteString(s)
-	return b.String()
-}
-
-func boolToString(b bool) string {
-	if b {
-		return "true"
+		return append(b, EmptyValue...)
+	default:
+		// TODO (CEV): consider panicking since this should be impossible
+		return append(b, EmptyValue...)
 	}
-	return "false"
 }
 
-// flagsToString checks all header flags and returns those
+// appendFlags checks all header flags and appends those
 // that are set as a string separated with commas
-func flagsToString(h dns.MsgHdr) string {
-	flags := make([]string, 7)
-	i := 0
-
+func appendFlags(b []byte, h dns.MsgHdr) []byte {
+	origLen := len(b)
 	if h.Response {
-		flags[i] = "qr"
-		i++
+		b = append(b, "qr,"...)
 	}
-
 	if h.Authoritative {
-		flags[i] = "aa"
-		i++
+		b = append(b, "aa,"...)
 	}
 	if h.Truncated {
-		flags[i] = "tc"
-		i++
+		b = append(b, "tc,"...)
 	}
 	if h.RecursionDesired {
-		flags[i] = "rd"
-		i++
+		b = append(b, "rd,"...)
 	}
 	if h.RecursionAvailable {
-		flags[i] = "ra"
-		i++
+		b = append(b, "ra,"...)
 	}
 	if h.Zero {
-		flags[i] = "z"
-		i++
+		b = append(b, "z,"...)
 	}
 	if h.AuthenticatedData {
-		flags[i] = "ad"
-		i++
+		b = append(b, "ad,"...)
 	}
 	if h.CheckingDisabled {
-		flags[i] = "cd"
-		i++
+		b = append(b, "cd,"...)
 	}
-	return strings.Join(flags[:i], ",")
+	if n := len(b); n > origLen {
+		return b[:n-1] // trim trailing ','
+	}
+	return b
 }
 
-// addrToRFC3986 will add brackets to the address if it is an IPv6 address.
-func addrToRFC3986(addr string) string {
-	if strings.Contains(addr, ":") {
-		return "[" + addr + "]"
+// appendAddrToRFC3986 will add brackets to the address if it is an IPv6 address.
+func appendAddrToRFC3986(b []byte, addr string) []byte {
+	if strings.IndexByte(addr, ':') != -1 {
+		b = append(b, '[')
+		b = append(b, addr...)
+		b = append(b, ']')
+	} else {
+		b = append(b, addr...)
 	}
-	return addr
+	return b
 }
 
+type nodeType int
+
 const (
-	headerReplacer = "{>"
-	labelReplacer  = "{/"
-	// EmptyValue is the default empty value.
-	EmptyValue = "-"
+	typeLabel    nodeType = iota // "{type}"
+	typeLiteral                  // "foo"
+	typeMetadata                 // "{/metadata}"
 )
+
+// A node represents a segment of a parsed format.  For example: "A {type}"
+// contains two nodes: "A " (literal); and "{type}" (label).
+type node struct {
+	value string // Literal value, label or metadata label
+	typ   nodeType
+}
+
+// A replacer is an ordered list of all the nodes in a format.
+type replacer []node
+
+func parseFormat(s string) replacer {
+	// Assume there is a literal between each label - its cheaper to over
+	// allocate once than allocate twice.
+	rep := make(replacer, 0, strings.Count(s, "{")*2)
+	for {
+		// We find the right bracket then backtrack to find the left bracket.
+		// This allows us to handle formats like: "{ {foo} }".
+		j := strings.IndexByte(s, '}')
+		if j < 0 {
+			break
+		}
+		i := strings.LastIndexByte(s[:j], '{')
+		if i < 0 {
+			// Handle: "A } {foo}"
+			//            ^j
```

Comment:
just `^`? Don't understand the "symbol" `^j`

Reply:
Will fix.